Exponential methods for future combinations.

hbandura opened the pull request.

Commit:
Exponential methods for future combinations.

thorvald commented on compiler/cpp/src/generate/t_java_generator.cc:
```diff
@@ -2579,7 +2580,11 @@ void t_java_generator::generate_service_future_interface(t_service* tservice) {
-    indent(f_service_) << function_signature_future(*f_iter) << ";" << endl << endl;
+    t_function* t_f = *f_iter;
+    int mask;
+    for (mask = 0; mask < (1 << t_f->get_arglist()->get_members().size()); mask++) {
```
Could we make this an option? If we want to push this upstream, I imagine it makes sense to let people choose between "only one function that only takes futures" and "the 2^n varations".

hbandura replied:
Sure. What would the options be? I imagine the all futures and also the all 'non-futures' make sense.

thorvald replied:
Good question. Let's start with "future-only" by default, and if you specify "mixed-future" or something like that, it generates all the variations?

thorvald commented on compiler/cpp/src/generate/t_java_generator.cc:
```diff
@@ -3967,14 +3972,42 @@ string t_java_generator::function_signature_async(t_function* tfunction, bool us
+ * modified to use a future of the argument type according to the mask
+ * given.
+ */
+string t_java_generator::futures_argument_list(t_struct* tstruct, int params_futures_mask) {
```
This function is almost an exact copy of argument_list. Can we modify argument_list instead, and have a default parameter at the end for futures_mask which defaults to 0?

hbandura replied:
This was on purpose. Since I don't know our plans for this library, I tried to keep all the medallia code separated as much as possible, to facilitate future merges in following versions. If the plan is to merge this to the original thrift library, then your suggestions makes sense.

thorvald commented on compiler/cpp/src/generate/t_java_generator.cc:
```diff
@@ -3967,14 +3972,42 @@ string t_java_generator::function_signature_async(t_function* tfunction, bool us
+    string type;
+    if (params_futures_mask & (1 << i)) {
+      // The mask indicates to use a future for this argument
+      type = "CompletableFuture<" + type_name((*f_iter)->get_type(), true) + ">";
```
This is a consumer, so let's consume the most generic thing we can; `Future<T>` instead of `CompletableFuture<T>`

There is one thing we should probably discuss here though; what do we do about numeric type promotion? For example, if the function expects `Future<Long>`, and we feed it a `Future<Integer>`, should that compile? Right now, it won't; to do that we'd either need even more overrides, or we'd have to accept a `Future<Number>`. `Future<Number>` has a different problem; you can now feed a long to a function that expects a short, and you'll get a runtime rather than compile-time error. I'm happy postponing this problem for now.

Edit: Add ` -- comment editing rocks.

hbandura replied:
1) I was going to ask the same thing, but since I hadn't yet dwelled into the client implementation, went with CF instead. 

2) I didn't follow the problem you are stating. Why won't it compile? a(Future<?> b) will compile against a(futureB). 

Future is also generically typed (Future<T>) so why would we have an issue at runtime? (Unless someone intentionally casts it)

hbandura replied:
Oh, I see that github removed all my generics <      T     >  , so that's probably what happened in your comment also. :)

thorvald replied:
AARGH. GitHub removes < and >, the comment makes no sense without them.

While a `Long` will accept a `Integer` in its place, a `Future<Long>` will not accept a `Future<Integer>` in its place.

``` java
    public void func(Future<Long> x) {
    }
    public void bad() {
        Future<Integer> v = null;
        func(v);
    }
```

Does not compile.

hbandura replied:
Ok, I now think I understood the problem. You were probably talking about something receiving Future< Long > and feeding it a Future< Integer > ? As you said, better avoid this now. We could later provide converting functions for number type promotion futures Futures.toLong(Future< Integer > intFuture); It's not ideal but I don't think it will be a super common use case.

thorvald replied:
To clarify original comment, it talked about Future< Long >, Future< Integer > and Future< Number >

thorvald commented:
Beyond comments, looks good to me.

Commit:
Code review updates: java:argument_list unified with java:futures_argument_list.
Using Future instead of CompletableFuture as FutureIFace arguments
Using mixed-futures as option to create all combinations on methods for FutureIFace (default is no futures).

thorvald commented on compiler/cpp/src/generate/t_java_generator.cc:
```diff
@@ -2579,7 +2584,15 @@ void t_java_generator::generate_service_future_interface(t_service* tservice) {
   for (f_iter = functions.begin(); f_iter != functions.end(); ++f_iter) {
-    indent(f_service_) << function_signature_future(*f_iter) << ";" << endl << endl;
+    t_function* t_f = *f_iter;
+    int mask;
```
Inline in for-loop

hbandura replied:
Question about this, I was going to inline it as "for (int mask ...)" but then I saw that f_iter is declared in a different line, and remembered about some compilers complaining about that inline definition, so I wasn't sure what the target compilers for thrift are. 100% safe to write the "for (int mask" idiom?

thorvald replied:
Yes, it's safe. If you want to be style-consistent, then leave "int i" on a separate line, then initialize it in the for-loop.

thorvald commented on compiler/cpp/src/generate/t_java_generator.cc:
```diff
@@ -3967,52 +3980,58 @@ string t_java_generator::function_signature_async(t_function* tfunction, bool us
   return "CompletableFuture<" + return_type + "> " + prefix + fn_name + "(" + arg_list + ")";
 }
 
-string t_java_generator::async_function_call_arglist(t_function* tfunc, bool use_base_method, bool include_types) {
```
Why was this moved?

hbandura replied:
Argh. Accident, I modified the future argument list which was in a different place, instead of the normal argument list. I'll move it back so it disappears form the diff.

Commit:
Code Review updates: move declaration of an int variable,
move function to previous place to minimize diff.

erdody commented on compiler/cpp/src/generate/t_java_generator.cc:
```diff
@@ -323,6 +326,7 @@ class t_java_generator : public t_oop_generator {
 
   // Generate FutureIface (based on CompletableFuture)
   bool future_iface_;
+  bool mixed_futures_;
```
Can you please add a comment about what the parameter does?
Also please add it to the command line help, below in this file.

hbandura replied:
Pull request updated with your suggestions.

Commit:
Review updates: Added comments and command-line documentation for mixed-futures parameter.
